File: Medium/ImplementTrie.js
class TrieNode {
    constructor() {
      this.children = {};
      this.endOfWord = false;
    }
  }
  
  var Trie = function() {
      this.root = new TrieNode();
  };
  
  /** 
   * @param {string} word
   * @return {void}
   */
  Trie.prototype.insert = function(word) {
      let current = this.root;
      
      for (let i = 0; i < word.length; i++) {
          let c = word.charAt(i);
          if (current[c] === undefined) {
              current[c] = new TrieNode();
          }
          current = current[c];
      }
      
      current.endOfWord = true;
  };
  
  /** 
   * @param {string} word
   * @return {boolean}
   */
  Trie.prototype.search = function(word) {
      let current = this.root;
      
      for (let i = 0; i < word.length; i++) {
          let c = word.charAt(i);
          if (current[c] === undefined) {
              return false;
          }
          current = current[c];
      }
      
      return current.endOfWord;
  };
  
  /** 
   * @param {string} prefix
   * @return {boolean}
   */
  Trie.prototype.startsWith = function(prefix) {
      
      let current = this.root;
      
      for (let i = 0; i < prefix.length; i++) {
          let c = prefix.charAt(i);
          if (current[c] === undefined) {
              return false;
          }
          current = current[c];
      }
      
      return true;
  };
  
  /** 
   * Your Trie object will be instantiated and called as such:
   * var obj = new Trie()
   * obj.insert(word)
   * var param_2 = obj.search(word)
   * var param_3 = obj.startsWith(prefix)
   */
  

File: C#/Medium/ImplementTrie.cs
/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.Insert(word);
 * bool param_2 = obj.Search(word);
 * bool param_3 = obj.StartsWith(prefix);
 */
public class Trie {

    private TrieNode _root;

    public Trie()
    {
        _root = new TrieNode();
    }
    
    public void Insert(string word)
    {
        TrieNode current = _root;
        for (int i = 0; i < word.Length; i++)
        {
            char c = word[i];
            if (!current.Dict.ContainsKey(c))
            {
                current.Dict.Add(c, new TrieNode());
            }

            current = current.Dict[c];
            if (i == word.Length - 1) current.IsWord = true;
        }
    }
    
    public bool Search(string word)
    {
        TrieNode current = _root;
        for (int i = 0; i < word.Length; i++)
        {
            char c = word[i];
            if (!current.Dict.ContainsKey(c)) return false;
            current = current.Dict[c];
            if (i == word.Length - 1) return current.IsWord;
        }

        return false;
    }
    
    public bool StartsWith(string prefix)
    {
        TrieNode current = _root;
        for (int i = 0; i < prefix.Length; i++)
        {
            char c = prefix[i];
            if (!current.Dict.ContainsKey(c)) return false;
            current = current.Dict[c];
        }

        return true;
    }
}

public class TrieNode
{
    public bool IsWord;
    public Dictionary<char, TrieNode> Dict;

    public TrieNode(bool isWord = false)
    {
        IsWord = isWord;
        Dict = new Dictionary<char, TrieNode>();
    }
}

/*

    Insert
    Time: O(n)
    Space: O(n)

    Search
    Time: O(n)
    Space: O(1)

    Prefix
    Time: O(n)
    Space: O(1)

    Where n is the length of the input string (word or prefix).

*/
